File: frontend/src/api/interface/cronjob.ts
import { ReqPage } from '.';

export namespace Cronjob {
    export interface Search extends ReqPage {
        info: string;
        groupIDs: Array<number>;
        orderBy?: string;
        order?: string;
    }
    export interface CronjobInfo {
        id: number;
        name: string;
        type: string;
        groupID: number;
        specCustom: boolean;
        spec: string;
        specs: Array<string>;
        specObjs: Array<SpecObj>;

        executor: string;
        isExecutorCustom: boolean;
        script: string;
        scriptMode: string;
        isCustom: boolean;
        command: string;
        inContainer: boolean;
        containerName: string;
        user: string;
        scriptID: number;
        appID: string;
        website: string;
        exclusionRules: string;
        ignoreFiles: Array<string>;
        dbType: string;
        dbName: string;
        url: string;
        isDir: boolean;
        files: Array<Item>;
        sourceDir: string;
        snapshotRule: snapshotRule;
        ignoreAppIDs: Array<Number>;
        withImage: boolean;

        websiteList: Array<string>;
        appIdList: Array<string>;
        dbNameList: Array<string>;

        sourceAccounts: Array<string>;
        downloadAccount: string;
        sourceAccountIDs: string;
        downloadAccountID: number;
        sourceAccountItems: Array<number>;

        retainCopies: number;
        ignoreErr: boolean;
        retryTimes: number;
        timeout: number;
        timeoutItem: number;
        timeoutUnit: string;
        status: string;
        secret: string;
        hasAlert: boolean;
        alertCount: number;
        alertTitle: string;
        alertMethod: string;
        alertMethodItems: Array<string>;
    }
    export interface Item {
        val: string;
    }
    export interface CronjobOperate {
        id: number;
        name: string;
        groupID: number;
        type: string;
        specCustom: boolean;
        spec: string;
        specs: Array<string>;
        specObjs: Array<SpecObj>;

        scriptID: number;
        appID: string;
        website: string;
        exclusionRules: string;
        dbType: string;
        dbName: string;
        url: string;
        isDir: boolean;
        sourceDir: string;
        snapshotRule: snapshotRule;

        //shell
        executor: string;
        scriptMode: string;
        script: string;
        command: string;
        containerName: string;
        user: string;

        sourceAccountIDs: string;
        downloadAccountID: number;
        retainCopies: number;
        retryTimes: number;
        timeout: number;
        ignoreErr: boolean;
        secret: string;

        alertCount: number;
        alertTitle: string;
        alertMethod: string;
    }
    export interface CronjobTrans {
        name: string;
        type: string;
        specCustom: boolean;
        spec: string;
        group: string;

        executor: string;
        scriptMode: string;
        script: string;
        command: string;
        containerName: string;
        user: string;
        url: string;

        scriptName: string;
        apps: Array<TransHelper>;
        websites: Array<string>;
        dbType: string;
        dbNames: Array<TransHelper>;

        exclusionRules: string;

        isDir: boolean;
        sourceDir: string;

        retainCopies: number;
        retryTimes: number;
        timeout: number;
        ignoreErr: boolean;
        snapshotRule: string;
        secret: string;

        sourceAccounts: Array<string>;
        downloadAccount: string;

        alertCount: number;
    }
    export interface TransHelper {
        name: string;
        detailName: string;
    }
    export interface snapshotTransHelper {
        withImage: boolean;
        ignoreApps: Array<TransHelper>;
    }
    export interface snapshotRule {
        withImage: boolean;
        ignoreAppIDs: Array<Number>;
    }
    export interface SpecObj {
        specType: string;
        week: number;
        day: number;
        hour: number;
        minute: number;
        second: number;
    }
    export interface CronjobDelete {
        ids: Array<number>;
        cleanData: boolean;
        cleanRemoteData: boolean;
    }
    export interface UpdateStatus {
        id: number;
        status: string;
    }
    export interface Download {
        recordID: number;
        backupAccountID: number;
    }
    export interface ScriptOptions {
        id: number;
        name: string;
        script: string;
    }
    export interface SearchRecord extends ReqPage {
        cronjobID: number;
        startTime: Date;
        endTime: Date;
        status: string;
    }
    export interface Record {
        id: number;
        taskID: string;
        file: string;
        startTime: string;
        records: string;
        status: string;
        message: string;
        targetPath: string;
        interval: number;
    }

    export interface ScriptInfo {
        id: number;
        name: string;
        script: string;
        groups: string;
        isInteractive: boolean;
        groupList: Array<number>;
        groupBelong: Array<string>;
        description: string;
        createdAt: Date;
    }
    export interface ScriptOperate {
        id: number;
        name: string;
        script: string;
        groups: string;
        description: string;
    }
}

File: agent/app/dto/cronjob.go
package dto

import (
	"time"
)

type PageCronjob struct {
	PageInfo
	Info     string `json:"info"`
	GroupIDs []uint `json:"groupIDs"`
	OrderBy  string `json:"orderBy" validate:"required,oneof=name status createdAt"`
	Order    string `json:"order" validate:"required,oneof=null ascending descending"`
}

type CronjobSpec struct {
	Spec string `json:"spec" validate:"required"`
}

type CronjobOperate struct {
	ID         uint   `json:"id"`
	Name       string `json:"name" validate:"required"`
	Type       string `json:"type" validate:"required"`
	GroupID    uint   `json:"groupID"`
	SpecCustom bool   `json:"specCustom"`
	Spec       string `json:"spec" validate:"required"`

	Executor      string `json:"executor"`
	ScriptMode    string `json:"scriptMode"`
	Script        string `json:"script"`
	Command       string `json:"command"`
	ContainerName string `json:"containerName"`
	User          string `json:"user"`

	ScriptID       uint         `json:"scriptID"`
	AppID          string       `json:"appID"`
	Website        string       `json:"website"`
	ExclusionRules string       `json:"exclusionRules"`
	DBType         string       `json:"dbType"`
	DBName         string       `json:"dbName"`
	URL            string       `json:"url"`
	IsDir          bool         `json:"isDir"`
	SourceDir      string       `json:"sourceDir"`
	SnapshotRule   SnapshotRule `json:"snapshotRule"`

	SourceAccountIDs  string `json:"sourceAccountIDs"`
	DownloadAccountID uint   `json:"downloadAccountID"`
	RetainCopies      int    `json:"retainCopies" validate:"number,min=1"`
	RetryTimes        int    `json:"retryTimes" validate:"number,min=0"`
	Timeout           uint   `json:"timeout" validate:"number,min=1"`
	IgnoreErr         bool   `json:"ignoreErr"`
	Secret            string `json:"secret"`

	AlertCount  uint   `json:"alertCount"`
	AlertTitle  string `json:"alertTitle"`
	AlertMethod string `json:"alertMethod"`

	CleanLogConfig
}

type CleanLogConfig struct {
	Scopes []string `json:"scopes"`
}

type SnapshotRule struct {
	WithImage    bool   `json:"withImage"`
	IgnoreAppIDs []uint `json:"ignoreAppIDs"`
}

type CronjobUpdateStatus struct {
	ID     uint   `json:"id" validate:"required"`
	Status string `json:"status" validate:"required"`
}

type CronjobClean struct {
	IsDelete        bool `json:"isDelete"`
	CleanData       bool `json:"cleanData"`
	CronjobID       uint `json:"cronjobID" validate:"required"`
	CleanRemoteData bool `json:"cleanRemoteData"`
}

type CronjobBatchDelete struct {
	CleanData       bool   `json:"cleanData"`
	CleanRemoteData bool   `json:"cleanRemoteData"`
	IDs             []uint `json:"ids" validate:"required"`
}

type CronjobInfo struct {
	ID         uint   `json:"id"`
	Name       string `json:"name"`
	Type       string `json:"type"`
	GroupID    uint   `json:"groupID"`
	SpecCustom bool   `json:"specCustom"`
	Spec       string `json:"spec"`

	Executor      string `json:"executor"`
	ScriptMode    string `json:"scriptMode"`
	Script        string `json:"script"`
	Command       string `json:"command"`
	ContainerName string `json:"containerName"`
	User          string `json:"user"`

	ScriptID       uint         `json:"scriptID"`
	AppID          string       `json:"appID"`
	Website        string       `json:"website"`
	ExclusionRules string       `json:"exclusionRules"`
	DBType         string       `json:"dbType"`
	DBName         string       `json:"dbName"`
	URL            string       `json:"url"`
	IsDir          bool         `json:"isDir"`
	SourceDir      string       `json:"sourceDir"`
	RetainCopies   int          `json:"retainCopies"`
	RetryTimes     int          `json:"retryTimes"`
	Timeout        uint         `json:"timeout"`
	IgnoreErr      bool         `json:"ignoreErr"`
	SnapshotRule   SnapshotRule `json:"snapshotRule"`

	SourceAccounts    []string `json:"sourceAccounts"`
	DownloadAccount   string   `json:"downloadAccount"`
	SourceAccountIDs  string   `json:"sourceAccountIDs"`
	DownloadAccountID uint     `json:"downloadAccountID"`

	LastRecordStatus string `json:"lastRecordStatus"`
	LastRecordTime   string `json:"lastRecordTime"`
	Status           string `json:"status"`
	Secret           string `json:"secret"`

	AlertCount uint `json:"alertCount"`
}

type CronjobImport struct {
	Cronjobs []CronjobTrans `json:"cronjobs"`
}
type CronjobTrans struct {
	Name       string `json:"name"`
	Type       string `json:"type"`
	GroupID    uint   `json:"groupID"`
	SpecCustom bool   `json:"specCustom"`
	Spec       string `json:"spec"`

	Executor      string `json:"executor"`
	ScriptMode    string `json:"scriptMode"`
	Script        string `json:"script"`
	Command       string `json:"command"`
	ContainerName string `json:"containerName"`
	User          string `json:"user"`
	URL           string `json:"url"`

	ScriptName string        `json:"scriptName"`
	Apps       []TransHelper `json:"apps"`
	Websites   []string      `json:"websites"`
	DBType     string        `json:"dbType"`
	DBNames    []TransHelper `json:"dbName"`

	ExclusionRules string `json:"exclusionRules"`

	IsDir     bool   `json:"isDir"`
	SourceDir string `json:"sourceDir"`

	RetainCopies uint64              `json:"retainCopies"`
	RetryTimes   uint                `json:"retryTimes"`
	Timeout      uint                `json:"timeout"`
	IgnoreErr    bool                `json:"ignoreErr"`
	SnapshotRule SnapshotTransHelper `json:"snapshotRule"`
	Secret       string              `json:"secret"`

	SourceAccounts  []string `json:"sourceAccounts"`
	DownloadAccount string   `json:"downloadAccount"`

	AlertCount  uint   `json:"alertCount"`
	AlertTitle  string `json:"alertTitle"`
	AlertMethod string `json:"alertMethod"`
}
type TransHelper struct {
	Name       string `json:"name"`
	DetailName string `json:"detailName"`
}
type SnapshotTransHelper struct {
	WithImage  bool          `json:"withImage"`
	IgnoreApps []TransHelper `json:"ignoreApps"`
}

type ScriptOptions struct {
	ID   uint   `json:"id"`
	Name string `json:"name"`
}

type SearchRecord struct {
	PageInfo
	CronjobID int       `json:"cronjobID"`
	StartTime time.Time `json:"startTime"`
	EndTime   time.Time `json:"endTime"`
	Status    string    `json:"status"`
}

type Record struct {
	ID         uint   `json:"id"`
	TaskID     string `json:"taskID"`
	StartTime  string `json:"startTime"`
	Records    string `json:"records"`
	Status     string `json:"status"`
	Message    string `json:"message"`
	TargetPath string `json:"targetPath"`
	Interval   int    `json:"interval"`
	File       string `json:"file"`
}
